Invalid token LIMIT in databse utils for android 11 +

## Issue explanation (write below this line)



## Exception
* __App Name:__ Amaze File Manager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.6.6
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 11 - 30
* __Device:__ umi
* __Model:__ Mi 10
* __Product:__ umi_eea
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:415)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.IllegalArgumentException: Invalid token LIMIT
at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:172)
at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:142)
at android.content.ContentProviderProxy.query(ContentProviderNative.java:472)
at android.content.ContentResolver.query(ContentResolver.java:1197)
at android.content.ContentResolver.query(ContentResolver.java:1128)
at android.content.ContentResolver.query(ContentResolver.java:1084)
at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.listRecentFiles(LoadFilesListTask.java:527)
at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:193)
at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:70)
at android.os.AsyncTask$3.call(AsyncTask.java:394)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 3 more

```
</details>
<hr>

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -28,8 +28,8 @@ android {
         applicationId "com.amaze.filemanager"
         minSdkVersion 14
         targetSdkVersion 30
-        versionCode 104
-        versionName "3.6.6"
+        versionCode 106
+        versionName "3.6.7"
         multiDexEnabled true
 
         vectorDrawables.useSupportLibrary = true

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -36,6 +36,7 @@
     <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
     <uses-permission android:name="android.permission.REQUEST_DELETE_PACKAGES" />
+    <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
 
     <uses-feature
         android:name="android.hardware.touchscreen"

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -37,6 +37,7 @@
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
     <uses-permission android:name="android.permission.REQUEST_DELETE_PACKAGES" />
     <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.QUERY_ALL_PACKAGES" />
 
     <uses-feature
         android:name="android.hardware.touchscreen"

--- a/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/LoadFilesListTask.java
+++ b/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/LoadFilesListTask.java
@@ -20,6 +20,9 @@
 
 package com.amaze.filemanager.asynchronous.asynctasks;
 
+import static android.os.Build.VERSION.SDK_INT;
+import static android.os.Build.VERSION_CODES.Q;
+
 import java.io.File;
 import java.lang.ref.WeakReference;
 import java.util.ArrayList;
@@ -51,10 +54,12 @@ import com.amaze.filemanager.utils.OnAsyncTaskFinished;
 import com.amaze.filemanager.utils.OnFileFound;
 import com.cloudrail.si.interfaces.CloudStorage;
 
+import android.content.ContentResolver;
 import android.content.Context;
 import android.database.Cursor;
 import android.net.Uri;
 import android.os.AsyncTask;
+import android.os.Bundle;
 import android.provider.MediaStore;
 import android.text.format.Formatter;
 import android.util.Log;
@@ -517,19 +522,37 @@ public class LoadFilesListTask
     }
 
     ArrayList<LayoutElementParcelable> recentFiles = new ArrayList<>(20);
-    final String[] projection = {MediaStore.Files.FileColumns.DATA};
+    final String[] projection = {
+      MediaStore.Files.FileColumns.DATA, MediaStore.Files.FileColumns.DATE_MODIFIED
+    };
     Calendar c = Calendar.getInstance();
     c.set(Calendar.DAY_OF_YEAR, c.get(Calendar.DAY_OF_YEAR) - 2);
     Date d = c.getTime();
-    Cursor cursor =
-        context
-            .getContentResolver()
-            .query(
-                MediaStore.Files.getContentUri("external"),
-                projection,
-                null,
-                null,
-                MediaStore.Files.FileColumns.DATE_MODIFIED + " DESC LIMIT 20");
+    Cursor cursor;
+    if (SDK_INT >= Q) {
+      Bundle queryArgs = new Bundle();
+      queryArgs.putInt(ContentResolver.QUERY_ARG_LIMIT, 20);
+      queryArgs.putStringArray(
+          ContentResolver.QUERY_ARG_SORT_COLUMNS,
+          new String[] {MediaStore.Files.FileColumns.DATE_MODIFIED});
+      queryArgs.putInt(
+          ContentResolver.QUERY_ARG_SORT_DIRECTION,
+          ContentResolver.QUERY_SORT_DIRECTION_DESCENDING);
+      cursor =
+          context
+              .getContentResolver()
+              .query(MediaStore.Files.getContentUri("external"), projection, queryArgs, null);
+    } else {
+      cursor =
+          context
+              .getContentResolver()
+              .query(
+                  MediaStore.Files.getContentUri("external"),
+                  projection,
+                  null,
+                  null,
+                  MediaStore.Files.FileColumns.DATE_MODIFIED + " DESC LIMIT 20");
+    }
     if (cursor == null) return recentFiles;
     if (cursor.getCount() > 0 && cursor.moveToFirst()) {
       do {

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -198,7 +198,8 @@ public class MainActivity extends PermissionsActivity
         SearchWorkerFragment.HelperCallbacks,
         CloudConnectionCallbacks,
         LoaderManager.LoaderCallbacks<Cursor>,
-        FolderChooserDialog.FolderCallback {
+        FolderChooserDialog.FolderCallback,
+        PermissionsActivity.OnPermissionGranted {
 
   private static final String TAG = TagsHelper.getTag(MainActivity.class);
 
@@ -469,43 +470,46 @@ public class MainActivity extends PermissionsActivity
     }
   }
 
-  private void checkForExternalPermission() {
+  @Override
+  public void onPermissionGranted() {
+    drawer.refreshDrawer();
+    TabFragment tabFragment = getTabFragment();
+    boolean b = getBoolean(PREFERENCE_NEED_TO_SET_HOME);
+    // reset home and current paths according to new storages
+    if (b) {
+      TabHandler tabHandler = TabHandler.getInstance();
+      tabHandler
+          .clear()
+          .subscribe(
+              () -> {
+                if (tabFragment != null) {
+                  tabFragment.refactorDrawerStorages(false);
+                  Fragment main = tabFragment.getFragmentAtIndex(0);
+                  if (main != null) ((MainFragment) main).updateTabWithDb(tabHandler.findTab(1));
+                  Fragment main1 = tabFragment.getFragmentAtIndex(1);
+                  if (main1 != null) ((MainFragment) main1).updateTabWithDb(tabHandler.findTab(2));
+                }
+                getPrefs().edit().putBoolean(PREFERENCE_NEED_TO_SET_HOME, false).commit();
+              });
+    } else {
+      // just refresh list
+      if (tabFragment != null) {
+        Fragment main = tabFragment.getFragmentAtIndex(0);
+        if (main != null) ((MainFragment) main).updateList();
+        Fragment main1 = tabFragment.getFragmentAtIndex(1);
+        if (main1 != null) ((MainFragment) main1).updateList();
+      }
+    }
+  }
 
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !checkStoragePermission()) {
-      requestStoragePermission(
-          () -> {
-            drawer.refreshDrawer();
-            TabFragment tabFragment = getTabFragment();
-            boolean b = getBoolean(PREFERENCE_NEED_TO_SET_HOME);
-            // reset home and current paths according to new storages
-            if (b) {
-              TabHandler tabHandler = TabHandler.getInstance();
-              tabHandler
-                  .clear()
-                  .subscribe(
-                      () -> {
-                        if (tabFragment != null) {
-                          tabFragment.refactorDrawerStorages(false);
-                          Fragment main = tabFragment.getFragmentAtIndex(0);
-                          if (main != null)
-                            ((MainFragment) main).updateTabWithDb(tabHandler.findTab(1));
-                          Fragment main1 = tabFragment.getFragmentAtIndex(1);
-                          if (main1 != null)
-                            ((MainFragment) main1).updateTabWithDb(tabHandler.findTab(2));
-                        }
-                        getPrefs().edit().putBoolean(PREFERENCE_NEED_TO_SET_HOME, false).commit();
-                      });
-            } else {
-              // just refresh list
-              if (tabFragment != null) {
-                Fragment main = tabFragment.getFragmentAtIndex(0);
-                if (main != null) ((MainFragment) main).updateList();
-                Fragment main1 = tabFragment.getFragmentAtIndex(1);
-                if (main1 != null) ((MainFragment) main1).updateList();
-              }
-            }
-          },
-          true);
+  private void checkForExternalPermission() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      if (!checkStoragePermission()) {
+        requestStoragePermission(this, true);
+      }
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+        requestAllFilesAccess(this);
+      }
     }
   }
 

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/superclasses/PermissionsActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/superclasses/PermissionsActivity.java
@@ -23,6 +23,7 @@ package com.amaze.filemanager.ui.activities.superclasses;
 import com.afollestad.materialdialogs.DialogAction;
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.amaze.filemanager.R;
+import com.amaze.filemanager.application.AppConfig;
 import com.amaze.filemanager.ui.dialogs.GeneralDialogCreation;
 import com.amaze.filemanager.utils.Utils;
 import com.google.android.material.snackbar.BaseTransientBottomBar;
@@ -33,6 +34,9 @@ import android.content.Intent;
 import android.content.pm.PackageManager;
 import android.net.Uri;
 import android.os.Build;
+import android.os.Environment;
+import android.provider.Settings;
+import android.util.Log;
 import android.widget.Toast;
 
 import androidx.annotation.NonNull;
@@ -42,14 +46,20 @@ import androidx.core.app.ActivityCompat;
 public class PermissionsActivity extends ThemedActivity
     implements ActivityCompat.OnRequestPermissionsResultCallback {
 
-  public static final int PERMISSION_LENGTH = 2;
-  public static final int STORAGE_PERMISSION = 0, INSTALL_APK_PERMISSION = 1;
+  private static final String TAG = PermissionsActivity.class.getSimpleName();
 
-  private OnPermissionGranted[] permissionCallbacks = new OnPermissionGranted[PERMISSION_LENGTH];
+  public static final int PERMISSION_LENGTH = 3;
+  public static final int STORAGE_PERMISSION = 0,
+      INSTALL_APK_PERMISSION = 1,
+      ALL_FILES_PERMISSION = 2;
+
+  private final OnPermissionGranted[] permissionCallbacks =
+      new OnPermissionGranted[PERMISSION_LENGTH];
 
   @Override
   public void onRequestPermissionsResult(
       int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults);
     if (requestCode == STORAGE_PERMISSION) {
       if (isGranted(grantResults)) {
         Utils.enableScreenRotation(this);
@@ -129,7 +139,7 @@ public class PermissionsActivity extends ThemedActivity
    * @param isInitialStart is the permission being requested for the first time in the application
    *     lifecycle
    */
-  public void requestPermission(
+  private void requestPermission(
       final String permission,
       final int code,
       @NonNull final MaterialDialog rationale,
@@ -162,6 +172,41 @@ public class PermissionsActivity extends ThemedActivity
                           android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                           Uri.parse(String.format("package:%s", getPackageName())))))
           .show();
+    }
+  }
+
+  /**
+   * Request all files access on android 11+
+   *
+   * @param onPermissionGranted permission granted callback
+   */
+  public void requestAllFilesAccess(@NonNull final OnPermissionGranted onPermissionGranted) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !Environment.isExternalStorageManager()) {
+      final MaterialDialog materialDialog =
+          GeneralDialogCreation.showBasicDialog(
+              this,
+              R.string.grant_all_files_permission,
+              R.string.grantper,
+              R.string.grant,
+              R.string.cancel);
+      materialDialog.getActionButton(DialogAction.NEGATIVE).setOnClickListener(v -> finish());
+      materialDialog
+          .getActionButton(DialogAction.POSITIVE)
+          .setOnClickListener(
+              v -> {
+                Utils.disableScreenRotation(this);
+                permissionCallbacks[ALL_FILES_PERMISSION] = onPermissionGranted;
+                try {
+                  Intent intent = new Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION);
+                  startActivity(intent);
+                } catch (Exception e) {
+                  Log.e(TAG, "Failed to initial activity to grant all files access", e);
+                  AppConfig.toast(this, getString(R.string.grantfailed));
+                }
+                materialDialog.dismiss();
+              });
+      materialDialog.setCancelable(false);
+      materialDialog.show();
     }
   }
 

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -729,5 +729,8 @@ You only need to do this once, until the next time you select a new location for
     <string name="select_by_date">Select by date</string>
     <string name="select_similar">Select similar</string>
     <string name="error_fetching_google_play_product_list">Error fetching product list from Google Play.</string>
+    <string name="grant_all_files_permission"><html><body>Amaze needs all files permission to continue.
+        \nOn next screen, please select <b>Amaze</b> and select <b>Allow access to manage all files</b> option.
+            \n\n<font color='#ff6347'><i>Canceling this dialog will exit the app.</i></font></body></html></string>
 </resources>
 

--- a/app/src/main/res/values/translators.xml
+++ b/app/src/main/res/values/translators.xml
@@ -42,7 +42,7 @@
 	<string name="vietnamese_translation_summary" translable="false" translatable="false">ngoisaosang</string>
     <string name="japanese_translation_summary" translatable="false">Naofumi Fukue</string>
     <string name="tamil_translation_summary" translatable="false">Kuralarasi for StarsSoft</string>
-    <string name="app_version" translatable="false">v3.6.6</string>
+    <string name="app_version" translatable="false">v3.6.7</string>
     <string name="author_1" translatable="false">Arpit Khurana</string>
     <string name="author_2" translatable="false">Vishal Nehra</string>
     <string name="developer_1" translatable="false">Emmanuel Messulam</string>

